openshift/dynamic/client.py

```diff
@@ -0,0 +1,387 @@
+            'namespaced_full': '/{}/namespaces/{{namespace}}/{}/{{name}}'.format(full_prefix, self.name)
+        }
+
+    def __getattr__(self, name):
+        return partial(getattr(self.client, name), self)
+
+
+class ResourceField(object):
+
+    def __init__(self, **kwargs):
+        self.__dict__.update(kwargs)
+
+    def __repr__(self):
+        return pformat(self.__dict__)
+
+    def __eq__(self, other):
+        return self.__dict__ == other.__dict__
+
+    def __getitem__(self, name):
+        return self.__dict__[name]
+
+    def __dir__(self):
+        return dir(type(self)) + list(self.__dict__.keys())
+
+
+class ResourceInstance(object):
+
+    def __init__(self, resource, instance):
+        self.resource_type = resource
+        self.attributes = self.__deserialize(instance)
+
+    def __deserialize(self, field):
+        if isinstance(field, dict):
+            return ResourceField(**{
+                k: self.__deserialize(v) for k, v in field.items()
+            })
+        elif isinstance(field, (list, tuple)):
+            return [self.__deserialize(item) for item in field]
+        else:
+            return field
+
+    def __serialize(self, field):
+        if isinstance(field, ResourceField):
+            return {
+                k: self.__serialize(v) for k, v in field.__dict__.items()
+            }
+        elif isinstance(field, (list, tuple)):
+            return [self.__serialize(item) for item in field]
+        else:
+            return field
+
+    def to_dict(self):
+        return self.__serialize(self.attributes)
+
+    def to_str(self):
+        return repr(self)
+
+    def __repr__(self):
+        return "ResourceInstance[{}]:\n  {}".format(
+            self.attributes.kind,
+            '  '.join(yaml.safe_dump(self.to_dict()).splitlines(True))
+        )
+
+    def __getattr__(self, name):
+        return getattr(self.attributes, name)
+
+    def __getitem__(self, name):
+        return self.attributes[name]
+
+    def __dir__(self):
+        return dir(type(self)) + list(self.attributes.__dict__.keys())
+
+
+class K8sResource(Resource):
+
+    @property
+    def prefix(self):
+        if self._prefix:
+            return self._prefix
+        elif self.apiversion == 'v1' and not self.group:
+            return '/api'
+        else:
+            return '/apis'
+
+
+class OpenshiftResource(Resource):
+
+    @property
+    def prefix(self):
+        if self._prefix:
+            return self._prefix
+        elif self.apiversion == 'v1' and not self.group:
+            return '/oapi'
+        else:
+            return '/apis'
+
+
+def flatten(l):
+    return [item for sublist in l for item in sublist]
+
+
+class DynamicClient(object):
+
+    def __init__(self, client):
+        self.client = client
+        self._groups = self.get_api_groups()
+        self._resources = flatten([self.get_resources_for_group(*group_parts) for group_parts in self._groups])
```

Comment:
I think this would make more sense as a `get_resources()` function that does all of this logic of getting all the resources for each group. Seems like it would be cleaner.